### PR TITLE
Support concurrent invocations of before class

### LINKED ISSUE #1035: @BeforeClass not executed in parallel when parallel="instances"
I personally came across the source of my problem when I found this stack overflow thread which helped me make sense of some strange behavior I've been encountering with a Factory + DataProvider + Parallel Instances

http://stackoverflow.com/questions/30027872/beforeclass-in-testng-is-not-executed-in-parallel-with-parallel-instances 

It doesn't look like a bug was ever filed, but I am getting burned with my selenium tests queueing against the grid, and waiting for a browser in the `@BeforeClass` annotation which sets up my test in a parent class. The child class has a `@Factory` to create all the instances up front, and the `@Test` method to actually carry out the test. The problem is that once a browser hits the grid queue, no other tests will start up, even though there are plenty of instances available (we test against various browsers, so having 1 browser flavor queue up doesn't mean that the grid is completely busy). 

The example posted in the stack overflow thread is:

``` java
@BeforeClass
public void beforeClass() throws InterruptedException{
    System.out.println("In Before Class ::: " + Thread.currentThread().getId());
    Thread.sleep(2000);
}


@Test
public void test(){
    System.out.println("In Test ::: " + Thread.currentThread().getId());
}

@AfterClass
public void afterClass(){
    System.out.println("In After Class ::: " + Thread.currentThread().getId());
}
```

I am running 6.9.9.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1035

Grouped all the “BeforeClass” related tests into
one class, so that its easy to add more when needed.

Fixes #1035  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/test/java/test/configuration/issue1035/MyFactory.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package test.configuration.issue1035;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.testng.annotations.Factory;
+
+public class MyFactory {
+  public static Set<InvocationTracker> tracker = Collections.newSetFromMap(new ConcurrentHashMap<>());
```
Comment (juherr) [inline src/test/java/test/configuration/issue1035/MyFactory.java]: Could be final and use the constant naming convention (upper case)

File: src/test/java/test/configuration/issue1035/InvocationTracker.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package test.configuration.issue1035;
+
+import java.util.Objects;
+
+public class InvocationTracker {
+
+  private long time;
```
Comment (juherr) [inline src/test/java/test/configuration/issue1035/InvocationTracker.java]: could be final

File: src/test/java/test/configuration/issue1035/TestclassExample.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  @Test
+  public void test() {
+  }
+
+  @AfterClass
+  public void afterClass() {
+  }
+
+  private void printer() {
```
Comment (juherr) [inline src/test/java/test/configuration/issue1035/TestclassExample.java]: could be static

**Review by juherr — COMMENTED** [review]
Few style remarks. Feel free to merge

File: src/test/java/test/configuration/issue1035/MyFactory.java (hunk 1 above)
Comment (krmahadevan) [reply src/test/java/test/configuration/issue1035/MyFactory.java]: fixed

File: src/test/java/test/configuration/issue1035/InvocationTracker.java (hunk 2 above)
Comment (krmahadevan) [reply src/test/java/test/configuration/issue1035/InvocationTracker.java]: fixed

File: src/test/java/test/configuration/issue1035/TestclassExample.java (hunk 3 above)
Comment (krmahadevan) [reply src/test/java/test/configuration/issue1035/TestclassExample.java]: This method needs access to `hashCode()` which is an instance method. So it cannot be static.

**Pull request merged**